### ISSUE TITLE
Surround (5.1 or 7.1) panning and spatial audio

### ISSUE THREAD

**Issue by dburongarcia** [WebAudio/web-audio-api#1479]
Hello, it seems PannerNode is hard-coded to just stereo output. I wonder if there is any other standard way to do panning and spatial audio for 5.1 or 7.1 systems. A surround panning system sorted like in Pro Tools would be great, allowing per-channel, spatial, divergence and LFE changes without having to code a custom solution for adapting each channel's gain.

**Comment by dburongarcia** [comment]
I've been doing some tests with spatial audio panning and webRTC. I connected the MediaStreamSource to a ChannelSplitterNode and then connected the first channel of the splitter (mic streaming is mono) to 6 different GainNodes that I connected separately to the 6 channels of a ChannelMerger.

It is working, although the lag between the original Stream and the 'spatial enabled' Stream is noticeable (500ms-1s) I tried then to connect the original Stream to multiple spatial Streams. On each one I notice a smaller lag (100-200ms) compared to the previous one.

I wonder what would be a realistic maximum number of this kind of solution: 'Source->Splitter->6xGain->Merger->Output' and if that could manage multiple Streams and Sources. I'm doing a graphical approximation for the Spatial audio to try to reduce the CPU usage, using the position of some points in a canvas to represent the speakers position and different geometries (circles, cones, etc...) with customizable gradients, so I just need to check the color. I know it's far from perfect, but at least it works.

Is this a 'proper way' to do spatial audio while the feature is not available? I also wonder why I always need to use the original StreamSource as the srcObject in a dummy audio tag to make the other 'Source->Splitter->6xGain->Merger->Output' Streams work when I use them as srcObject in other audio tags?

**Comment by hoch** [comment]
> It is working, although the lag between the original Stream and the 'spatial enabled' Stream is noticeable (500ms-1s) I tried then to connect the original Stream to multiple spatial Streams. On each one I notice a smaller lag (100-200ms) compared to the previous one.

This seems like an implementation issue. Perhaps file a bug toward the browser you're using?

> Is this a 'proper way' to do spatial audio while the feature is not available? I also wonder why I always need to use the original StreamSource as the srcObject in a dummy audio tag to make the other 'Source->Splitter->6xGain->Merger->Output' Streams work when I use them as srcObject in other audio tags?

I am not sure if this is the right venue for QnA. Also I wasn't really able to understand the question...

**Comment by padenot** [comment]
> I've been doing some tests with spatial audio panning and webRTC. I connected the MediaStreamSource to a ChannelSplitterNode and then connected the first channel of the splitter (mic streaming is mono) to 6 different GainNodes that I connected separately to the 6 channels of a ChannelMerger.
> 
> It is working, although the lag between the original Stream and the 'spatial enabled' Stream is noticeable (500ms-1s) I tried then to connect the original Stream to multiple spatial Streams. On each one I notice a smaller lag (100-200ms) compared to the previous one.

This is an implementation bug, to be filed in the implementation bug tracker.

> I wonder what would be a realistic maximum number of this kind of solution: 'Source->Splitter->6xGain->Merger->Output' and if that could manage multiple Streams and Sources. I'm doing a graphical approximation for the Spatial audio to try to reduce the CPU usage, using the position of some points in a canvas to represent the speakers position and different geometries (circles, cones, etc...) with customizable gradients, so I just need to check the color. I know it's far from perfect, but at least it works.

It depends on the implementation. Again, this is a not a spec concern. The spec tells you however,  that implementations MUST support at least 32 channels for an `AudioBuffer`.

> Is this a 'proper way' to do spatial audio while the feature is not available? I also wonder why I always need to use the original StreamSource as the srcObject in a dummy audio tag to make the other 'Source->Splitter->6xGain->Merger->Output' Streams work when I use them as srcObject in other audio tags?

This is the proper way (for now). What you see are probably implementation bugs that might or might not also be present in other implementations.

You can file an issue at for Chrome at <https://new.crbug.com> and for Firefox at <https://bugzilla.mozilla.org/enter_bug.cgi?product=Core&component=Web%20Audio>, where one of us will hopefully follow up.

**Comment by dburongarcia** [comment]
Yes, it is definitely an implementation issue in Chrome 64.0.3282.167 (Build oficial) (32 bits and 64 bits) In Firefox is working flawlessly without any lag. On the other hand, enumerateDevices is not working with output devices yet in Firefox... It's not a WebRTC issue because it happens also with normal MediaSources and not only Streams. I will file the issue for Chrome.

**Comment by rtoy** [comment]
File an issue against Chrome at crbug.com/new.  Be sure to include a simple repro case.  After creating the issue set the Component to Blink>WebAudio if you can.

**Comment by carstenschwede** [comment]
@dburongarcia Could you provide the code that exhibits the issue?
